pyexchange/gemini.py

```diff
@@ -0,0 +1,215 @@
+                       timestamp=int(item["timestamp"]))
+
+
+class GeminiApi(PyexAPI):
+  """ Gemini Exchange API interface
+
+  Implemented based on https://docs.gemini.com/rest-api/?python
+  """
+
+  logger = logging.getLogger()
+  
+  def __init__(self, api_server: str, api_key: str, api_secret: str, timeout: float) -> None:
+    assert(isinstance(api_server, str))
+    assert(isinstance(api_key, str))
+    assert(isinstance(api_secret, str))
+    assert(isinstance(timeout, float))
+
+    self.api_server = api_server
+    self.api_key = api_key
+    self.api_secret = api_secret
+    self.timeout = timeout
+
+  def get_balances(self):
+    balances = self._http_authenticated("POST", "/v1/balances")
+
+    return {
+      balance['currency']: {
+        "amount": balance["amount"],
+        "availableForTrade": balance["available"],
+        "availableForWithdrawal": balance["availableForWithdrawal"]
+      }
+      for balance in balances
+    }
+
+  def get_balance(self, coin: str):
+    assert(isinstance(coin, str))
+
+    balances = self.get_balances()
+    return balances.get(coin.upper())
+    
+  def get_orders(self, pair: str) -> List[GeminiOrder]:
+    assert(isinstance(pair, str))
+    pair = self._fix_pair(pair)
+
+    orders = self._http_authenticated("POST", "/v1/orders")
+    return list(map(lambda order: GeminiOrder.create(order), orders))
+  
+  def place_order(self, pair: str, is_sell: bool, price: Wad, amount: Wad) -> str:
+    assert(isinstance(pair, str))
+    assert(isinstance(is_sell, bool))
+    assert(isinstance(price, Wad))
+    assert(isinstance(amount, Wad))
+
+    pair = self._fix_pair(pair)
+
+    params = {
+      "symbol": pair,
+      "amount": str(amount),
+      "price": str(price),
+      "side": "sell" if is_sell else "buy",
+      "type": "exchange limit"
+    }
+    placed_order = self._http_authenticated("POST", "/v1/order/new", params)
+
+    return placed_order["order_id"]
+
+  def cancel_order(self, order_id: str, pair: str) -> bool:
+    assert(isinstance(order_id, str))
+    self.logger.info(f"Cancelling order #{order_id} on pair {pair}...")
+
+    canceled_order = self._http_authenticated("POST", "/v1/order/cancel", {"order_id": order_id})
+
+    return canceled_order["is_cancelled"]
+
+  def get_trades(self, pair: str, page_number: int = 1) -> List[GeminiTrade]:
+    assert(isinstance(pair, str))
+    assert(isinstance(page_number, int))
+
+    trades = self._http_authenticated("POST", "/v1/mytrades", {"symbol": self._fix_pair(pair)})
+    return list(map(lambda trade: GeminiTrade.from_my_trade(pair, trade), trades))
+
+  def get_all_trades(self, pair: str, page_number: int = 1) -> List[GeminiTrade]:
+    assert(isinstance(pair, str))
+    assert(isinstance(page_number, int))
+    
+    all_trades = self._http_unauthenticated("GET", f"/v1/trades/{self._fix_pair(pair)}")
+
+    return list(map(lambda trade: GeminiTrade.from_trade(pair, trade), all_trades))
```

Comment:
Does all trades endpoint paginate?

Reply:
No, it returns up to 500 trades up to 7 days ago. `since` timestamp can be also defined

https://docs.gemini.com/rest-api/#trade-history